### PR TITLE
Ticket 115: Verifying inclusion proofs

### PR DESCRIPTION
**Pull request by katjoyce**
- Make 'opensource' inclusion-proof-checking function more closely resemble
  the actual opensource inclusion-proof-checking function
- Add more thorough testing of RFC algorithm for verifying inclusion
  proofs
- Fix incorrect results when leaf_index >= tree_size
- Fix incorrect results when tree height is too large
- Reflect these fixes in the RFC

### CODE REVIEW

File: test_rfc_algorithms.py [hunk 1]
```diff
@@ -193,7 +198,11 @@ def check_inclusion_via_rfc_algorithm(hash, leaf_index, audit_path, tree_size, r
     fn >>= 1
     sn >>= 1
 
-  # 4.  Compare "r" against the "root_hash".  If they are equal,
+  # 5.  If "sn" is greater than 0, fail the proof verification.
+  if sn:
```
Comment (eranmes) [inline test_rfc_algorithms.py]: Nit: You can merge 5 & 6 by returning (sn == 0) and (r == root_hash)

File: rfc6962-bis.xml [hunk 2]
```diff
@@ -1433,6 +1434,7 @@ but it is expected there will be a variety.
                 <vspace blankLines="1" />
                 Finally, right-shift both <spanx style="verb">fn</spanx> and <spanx style="verb">sn</spanx> one time.
               </t>
+              <t>If <spanx style="verb">sn</spanx> is greater than 0, fail the proof verification.</t>
```
Comment (eranmes) [inline rfc6962-bis.xml]: See comment about merging this check with the check later on in the .py file.

**Comment by eranmes** [comment]
Overall looks good - thanks for thoroughly examining the inclusion proof checking algorithm, without it we would have published the wrong algorithm in the RFC.
One minor comment.

File: test_rfc_algorithms.py (hunk 1 above)
Comment (katjoyce) [inline test_rfc_algorithms.py]: Done

File: rfc6962-bis.xml (hunk 2 above)
Comment (katjoyce) [inline rfc6962-bis.xml]: Done

File: test_rfc_algorithms.py [hunk 3]
```diff
@@ -254,30 +261,40 @@ def cross_check_consistency_against_opensource_algorithm(first, second, first_ha
 
 
 def cross_check_inclusion_via_opensource(hash, leaf_index, audit_path, tree_size, root_hash):
-  audit_path = audit_path[:]
+  path_root_hash = opensource_root_from_path(leaf_index, tree_size, audit_path, hash)
+  if path_root_hash is "":
```
Comment (benlaurie) [inline test_rfc_algorithms.py]: ```if not path_root_hash:``` is more pythonic, even though it pains me.

File: test_rfc_algorithms.py [hunk 4]
```diff
@@ -254,30 +261,40 @@ def cross_check_consistency_against_opensource_algorithm(first, second, first_ha
...
+    return False
+  return path_root_hash == root_hash
+
+def opensource_root_from_path(leaf_index, tree_size, audit_path, hash):
+  if (leaf_index >= tree_size or leaf_index < 0):
```
Comment (benlaurie) [inline test_rfc_algorithms.py]: No ().

**Comment by eranmes** [comment]
LGTM once Ben's comments are addressed.

File: test_rfc_algorithms.py (hunk 3 above)
Comment (katjoyce) [inline test_rfc_algorithms.py]: Done

File: test_rfc_algorithms.py (hunk 4 above)
Comment (katjoyce) [inline test_rfc_algorithms.py]: Done

**Comment by benlaurie** [comment]
Splendid. I believe there's also a trac ticket needs updating...

**Pull request merged**